contracts/StorageManager.sol

```diff
@@ -172,177 +156,200 @@ contract StorageManager {
-        agreement.numberOfPeriodsWithdrawn = 0;
-        //NO_OVERFLOW reasoning: REF_CAPACITY
-        offerRegistry[msg.sender].availableCapacity = offerRegistry[msg.sender].availableCapacity + agreement.size;
-        agreement.startDate = 0;
-        //NO_OVERFLOW reasoning:  verified in "agreement expired" require and see REF_MAX_TRANSFER
-        msg.sender.transfer(agreement.numberOfPeriodsDeposited - periodsPast * agreement.billingPrice);
-        emit AgreementStopped(agreementReference);
+    function withdrawFunds(bytes32[] memory dataReference, address provider, uint256 amount) public payable {
+        Offer storage offer = offerRegistry[provider];
+        require(offer.availableCapacity != 0, "StorageManager: Offer for this Provider doesn't exist");
+
+        bytes32 agreementReference = getAgreementReference(dataReference, msg.sender);
+        Agreement storage agreement = offer.agreementRegistry[agreementReference];
+        require(agreement.size != 0, "StorageManager: Agreement for this Offer doesn't exist");
+
+        uint256 maxWithdrawableFunds;
+        if (agreement.lastPayoutDate == 0) {
+            // Agreement is inactive, consumer can withdraw all funds
+            maxWithdrawableFunds = agreement.availableFunds;
+        } else {
+            // Consumer can withdraw all funds except for those already used for past storage hosting
+            // AND for current period
+            maxWithdrawableFunds = agreement.availableFunds - _calculateSpentFunds(agreement) - (agreement.billingPrice * agreement.size);
+        }
+
+        if (amount == 0) {
+            amount = maxWithdrawableFunds;
+        }
+
+        require(amount <= maxWithdrawableFunds, "StorageManager: Amount is too big");
+        agreement.availableFunds = agreement.availableFunds.sub(amount);
+        msg.sender.transfer(amount);
+        emit AgreementFundsWithdrawn(agreementReference, amount);
     }
 
     /**
+    >> FOR PROVIDER
     @notice payout already earned funds of one or more Agreement
     @dev 
-    - any safeMath operations in this function don't cause a deadlock, as a possible revert just means we have to add less agreementReferences
-    - Provider must call an expired agreement themselves as soon as the agreement is expired, to add back the capacity.
+    - Provider must call an expired agreement themselves as soon as the agreement is expired, to add back the capacity to their Offer.
     @param agreementReferences reference to one or more Agreement
     */
     function payoutFunds(bytes32[] memory agreementReferences) public {
+        _payoutFunds(agreementReferences, msg.sender);
+    }
+
+    function _payoutFunds(bytes32[] memory agreementReferences, address payable provider) internal {
+        Offer storage offer = offerRegistry[provider];
+        require(offer.availableCapacity != 0, "StorageManager: Offer for this Provider doesn't exist");
         uint256 toTransfer;
```

Comment:
The default value is 0, but I prefer to set it to 0 
:) 
uint256 toTransfer = 0;